Adding stronger password security

jloveland opened the pull request:
Adding stronger password requirements for improving security based on OWASP. Fixing #848

mleanos commented on modules/users/client/directives/password-validator.client.directive.js:
```diff
@@ -0,0 +1,42 @@
+      require: 'ngModel',
+      link: function(scope, element, attrs, modelCtrl) {
+        modelCtrl.$parsers.unshift(function (password) {
+          var owaspPasswordStrengthTest = $window.owaspPasswordStrengthTest;
```
Can we have a client service for `owasp`? Some user's might want to extend it, or have a bit more control over it's behavior.

mleanos commented:
@jloveland This all looks pretty good. I left a line comment. I haven't taken this for a test drive yet, but I plan to sometime tonight. Overall, I like what I see. Great job!

jloveland commented:
Ok, I will add a client service tonight. 

> On Sep 3, 2015, at 12:16 AM, Michael Leanos <notifications@github.com> wrote:
> 
> @jloveland This all looks pretty good. I left a line comment. I haven't taken this for a test drive yet, but I plan to sometime tonight. Overall, I like what I see. Great job!
> 
> —
> Reply to this email directly or view it on GitHub.
>

jloveland commented:
ok, added a client service, check it out.

mleanos commented:
@jloveland I'll look at this again tonight. Got side-tracked yesterday with the jshint issue.

mleanos commented on modules/users/tests/server/user.server.model.tests.js:
```diff
@@ -217,9 +216,87 @@ describe('User Model Unit Tests:', function () {
+
+    it('should not allow more than 3 or more repeating characters ', function (done) {
+      var _user1 = new User(user1);
+      _user1.password = 'L0veSexxxSecre+God';
```
might want to change that :)

jloveland replied:
haha..yeah, I copied the test passwords from the owasp lib [here](https://github.com/nowsecure/owasp-password-strength-test/blob/master/test.js#L30), I can change that to be less *interesting*..

mleanos commented:
@jloveland I took this for a test drive, and everything is working perfectly. I see you added tests as well :) 

I left a line comment about one of the passwords you have set in one of the new tests. The only other suggestion might be to remove the pop-over when the validation is at 100%. This is merely a suggestion; I just wanted to point it out.

Everything else looks really good. I think this is good to go. Thank you for knocking this out. It's really a great enhancement.

jloveland commented:
I can also disable the pop-over once the test passes.

jloveland commented:
this is still WIP. I noticed that
1. I missed the reset password form.
2. password reset request fails because of `undefined` password
So, I am working on these now.

I need some advice on how to implement closing the popover once the password field is valid. I tried `popover-is-open="{{userForm.password.$error}}"` and that was causing conflicts..

jloveland commented:
I added password validation and verification, fixed tests, etc. I did not implement disabling the popover, maybe that can be done on a different PR? 

Also, as you can see, I had some issues with squashing my commits properly, so I pushed anyway. Maybe someone can help me with the rebase?

simison commented:
Great! Testing this out. 

~~UX wise it would be clearer if those requirements would be visible already from the beginning, not only when I hit "sign up". Now it gives blue 80% (=good?) strength for complicated but short passwords.~~

__Edit:__ Oh, there indeed is a popup telling all that. It just doesn't appear until I type something into the field and then click out and focus again to the input.

Incidentally I have an easy to use Angular module for this stuff. :-) https://github.com/trustroots/trustpass Check if you like it.

jloveland commented:
I just squashed my commits, let me know what you think. @simison thanks for the suggestion. maybe we can evaluate your library and refactor with another PR once this is merged?

simison commented:
@jloveland sure! I'd try to fix that modal not appearing on first focus anyway. I think I've had that before too when trying to add content dynamically into popups/tooltips. Make sure you have the latest bootstrap-ui though.

mleanos commented:
@jloveland I've tested this out again, with the password recovery too. Looks good to me. Do you still need help with the rebasing?

To get my PR branches rebased, I usually get my local master branch current, with upstream (meanjs), and then checkout my PR branch, and perform `git rebase master` Sometimes, I'll get a warning that my origin & local have diverged. In which case, I just push to my origin with `--force`

jloveland commented:
@mleanos I think I got the rebase worked out. please let me know if something looks wrong.

@simison I saw a lot of references to similar issues when researching the *dynamic* popover. I could solve the popover issue by reverting to *static* popover...placing the message in the [html](https://github.com/jloveland/mean/blob/owasp-password-strength/modules/users/client/views/authentication/signup.client.view.html#L37) like so:
`popover="Please enter a passphrase or password with greater than 10 characters, numbers, lowercase, uppercase, and special characters."`

instead of the dynamic approach:
`popover="{{popoverMsg}}"`

This is not ideal, but it would fix the problem. What do you think?

mleanos commented:
@jloveland I'm not an expert with Git, so correct me if I'm wrong. I think having those other commits (other than from you) is the result of not squashing after your rebase. I usually try get my PR's down to one commit.

jloveland commented:
@mleanos, you were correct, I needed to `git rebase master` then squash commits.

Are there any Angular.js experts out there that can advise me on how to solve the issue with the dynamic popover message not displaying on focus properly? Otherwise, my next change will be to revert to static popover defined in the html.

simison commented:
It has something to do with dynamic content; try to make sure that variable always has some non-empty string content or try fiddling with `tooltip-enable="true|false"` or `tooltip-is-open="true|false"`.

jloveland commented:
- fixed the dynamic popover message scope issues by moving the message variable the service.
- made popover disabled when password strength passes
- made button sizes consistent with the rest of the views

mleanos commented:
@jloveland Very good. Glad you got those issues worked out. I've tested this, and everything is working wonderfully! Thanks. 

I think this is ready to go. @simison How's it look to you?

codydaig commented:
LGTM 

Thanks @jloveland! :-D

lirantal commented:
Generally reviewed it, LGTM.

mleanos commented:
@lirantal I had the same concern over existing user's. I tested this, and these changes didn't have an impact on my existing users.

These changes merely enforce the new password requirements, when a user attempts to change their password; or when a new user signs up.

@jloveland ?

jloveland commented:
@lirantal and @mleanos this does not affect existing users until they change their password. It will have immediate affect on users who signup.

lirantal commented:
Cool, let's continue with this then.

simison commented on modules/users/client/directives/password-verify.client.directive.js:
```diff
@@ -0,0 +1,33 @@
+angular.module('users')
+  .directive("passwordVerify", function() {
+    return {
+      require: "ngModel",
```
@jloveland JSHint is complaining about not using single quotes here. Could you fix and send another PL?

ilanbiala replied:
also in the directive name above?

simison commented on modules/users/client/directives/password-verify.client.directive.js:
```diff
@@ -0,0 +1,33 @@
+                modelCtrl.$setValidity("passwordVerify", false);
+                return undefined;
+              } else {
+                modelCtrl.$setValidity("passwordVerify", true);
```
should be single quotes

simison commented on modules/users/client/services/password-validator.client.service.js:
```diff
@@ -0,0 +1,19 @@
+        return result;
+      },
+      getPopoverMsg: function () {
+        var popoverMsg = "Please enter a passphrase or password with greater than 10 characters, numbers, lowercase, upppercase, and special characters.";
```
should be single quotes

jloveland replied:
Thanks @simison , I didn't see this show after running `grunt jshint` on that branch. Output from the command:
```
grunt jshint
Running "jshint:all" (jshint) task
>> 105 files lint free.

Done, without errors.
```
Any thoughts on why I don't get the errors?
I can submit another PR to correct this though...

simison replied:
A good question. I just copied the directive to my meanjs-based project and got this. Is /client/services/ a new folder?

BTW Thanks for writing these!



> On 10 Sep 2015, at 14:21, Jason Loveland <notifications@github.com> wrote:
> 
> In modules/users/client/services/password-validator.client.service.js:
> 
> > @@ -0,0 +1,19 @@
> > +'use strict';
> > +
> > +// PasswordValidator service used for testing the password strength
> > +angular.module('users').factory('PasswordValidator', ['$window',
> > +  function ($window) {
> > +    var owaspPasswordStrengthTest = $window.owaspPasswordStrengthTest;
> > +
> > +    return {
> > +      getResult: function (password) {
> > +        var result = owaspPasswordStrengthTest.test(password);
> > +        return result;
> > +      },
> > +      getPopoverMsg: function () {
> > +        var popoverMsg = "Please enter a passphrase or password with greater than 10 characters, numbers, lowercase, upppercase, and special characters.";
> 
> Thanks @simison , I didn't see this show after running grunt jshint on that branch. Output from the command:
> 
> grunt jshint
> Running "jshint:all" (jshint) task
> >> 105 files lint free.
> 
> Done, without errors.
> 
> Any thoughts on why I don't get the errors?
> I can submit another PR to correct this though...
> 
> —
> Reply to this email directly or view it on GitHub.
>
